exonum-java-binding/testkit/src/main/java/com/exonum/binding/testkit/TestKit.java

```diff
@@ -448,39 +433,64 @@ public Builder withValidators(short validatorCount) {
     }
 
     /**
-     * Adds a service with which the TestKit would be instantiated. Several services can be added.
+     * Adds a service artifact which would be deployed by the TestKit. Several service artifacts
+     * can be added.
+     *
+     * <p>Once the service artifact is deployed, the service instances can be added with
+     * {@link #withService(ServiceArtifactId, String, int, MessageLite)}.
      */
-    public Builder withService(Class<? extends ServiceModule> serviceModule) {
-      services.add(serviceModule);
+    public Builder withDeployedArtifact(ServiceArtifactId serviceArtifactId, String artifactFilename) {
+      serviceArtifactFilenames.put(serviceArtifactId, artifactFilename);
       return this;
     }
 
     /**
-     * Adds services with which the TestKit would be instantiated.
+     * Adds a service specification with which the TestKit would create the corresponding service
+     * instance. Several service specifications can be added. All services are started and
+     * configured before the genesis block.
+     *
+     * <p>Note that the corresponding service artifact with equal serviceArtifactId should be
+     * deployed with {@link #withDeployedArtifact(ServiceArtifactId, String)}.
      */
-    @SafeVarargs
-    public final Builder withServices(Class<? extends ServiceModule> serviceModule,
-                                      Class<? extends ServiceModule>... serviceModules) {
-      return withServices(Lists.asList(serviceModule, serviceModules));
+    public Builder withService(ServiceArtifactId serviceArtifactId, String serviceName,
+                               int serviceId, MessageLite configuration) {
+      checkServiceId(serviceId);
+      checkServiceArtifactIsDeployed(serviceArtifactId);
+      ServiceSpec serviceSpec = new ServiceSpec(serviceName, serviceId,
+          configuration.toByteArray());
+      services.put(serviceArtifactId, serviceSpec);
+      return this;
+    }
+
+    private void checkServiceArtifactIsDeployed(ServiceArtifactId serviceArtifactId) {
+      checkArgument(serviceArtifactFilenames.containsKey(serviceArtifactId),
+          "Service %s should be deployed first in order to be created", serviceArtifactId);
     }
 
     /**
-     * Adds services with which the TestKit would be instantiated.
+     * Adds a service specification with which the TestKit would create the corresponding service
+     * instance with default configuration. Several service specifications can be added. All
```

Comment:
with no configuration. (there is no 'default' in the core, it is default of the testkit).